esp8266/modesp.c

```diff
@@ -35,6 +35,492 @@
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(esp_espconn_ondisconnect_obj, esp_espconn_ondisconnect);
+
+STATIC const mp_map_elem_t esp_espconn_locals_dict_table[] = {
+    { MP_OBJ_NEW_QSTR(MP_QSTR___del__), (mp_obj_t)&esp_espconn___del___obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_close),   (mp_obj_t)&esp_espconn_close_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_bind),    (mp_obj_t)&esp_espconn_bind_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_listen),  (mp_obj_t)&esp_espconn_listen_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_accept),  (mp_obj_t)&esp_espconn_accept_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_connect), (mp_obj_t)&esp_espconn_connect_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_send),    (mp_obj_t)&esp_espconn_send_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_recv),    (mp_obj_t)&esp_espconn_recv_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_sendto),  (mp_obj_t)&esp_espconn_sendto_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_recvfrom),    (mp_obj_t)&esp_espconn_recvfrom_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_onconnect),   (mp_obj_t)&esp_espconn_onconnect_obj },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_onrecv),      (mp_obj_t)&esp_espconn_onrecv_obj  },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_ondisconnect),(mp_obj_t)&esp_espconn_ondisconnect_obj },
+};
+STATIC MP_DEFINE_CONST_DICT(esp_espconn_locals_dict, esp_espconn_locals_dict_table);
+
+STATIC const mp_obj_type_t esp_espconn_type = {
+    { &mp_type_type },
+    .name = MP_QSTR_espconn,
+    .make_new = esp_espconn_make_new,
+    .locals_dict = (mp_obj_t)&esp_espconn_locals_dict,
+};
+
+typedef struct _esp_getaddrinfo_cb_struct_t {
+    mp_obj_t lambda;
+    mp_uint_t port;
+} esp_getaddrinfo_cb_struct_t;
+
+STATIC void esp_getaddrinfo_cb(const char *name, ip_addr_t *ipaddr,
+    void *arg) {
+    esp_getaddrinfo_cb_struct_t *vals = arg;
+    if (ipaddr != NULL) {
+        uint8_t ip[4];
+        ip[0] = (ipaddr->addr >> 24) & 0xff;
+        ip[1] = (ipaddr->addr >> 16) & 0xff;
+        ip[2] = (ipaddr->addr >>  8) & 0xff;
+        ip[3] = (ipaddr->addr >>  0) & 0xff;
+
+        mp_obj_tuple_t *tuple = mp_obj_new_tuple(5, NULL);
+
+        tuple->items[0] = MP_OBJ_NEW_SMALL_INT(0);
+        tuple->items[1] = MP_OBJ_NEW_SMALL_INT(0);
+        tuple->items[2] = MP_OBJ_NEW_SMALL_INT(0);
+        tuple->items[3] = MP_OBJ_NEW_QSTR(MP_QSTR_);
+        tuple->items[4] = esp_espconn_format_inet_addr(ip, vals->port);
+        call_function_1_protected(vals->lambda, tuple);
+    } else {
+        call_function_1_protected(vals->lambda, mp_const_none);
+    }
+    gc_free(vals);
+}
+
+STATIC mp_obj_t esp_getaddrinfo(mp_obj_t host_in, mp_obj_t port_in,
+    mp_obj_t lambda_in) {
+    mp_uint_t hlen;
+    const char *host = mp_obj_str_get_data(host_in, &hlen);
+    ip_addr_t ipaddr;
+
+    // Altough esp_espconn_gethostname specifies esp_espconn * as its first argument,
+    // it does not look like it actually uses it for anything so we are
+    // free to pass whatever we want
+    esp_getaddrinfo_cb_struct_t *arg =
+        gc_alloc(sizeof(esp_getaddrinfo_cb_struct_t), false);
```

Comment:
I think this is a dangerous use of malloc: we don't know where the pointer to arg is stored when you pass it to espconn_gethostbyname, so we don't know if the GC will scan it or not.  Best probably to just use some static bss memory for this and just tell users that getaddrinfo can't be called again without the first one returning.